File: backend/src/main/java/com/hardware/erp/tenant/service/SubscriptionCouponService.java
package com.hardware.erp.tenant.service;

import com.hardware.erp.common.dto.PageResponse;
import com.hardware.erp.tenant.dto.SubscriptionCouponRedemptionResponse;
import com.hardware.erp.tenant.dto.SubscriptionCouponRequest;
import com.hardware.erp.tenant.dto.SubscriptionCouponResponse;
import com.hardware.erp.tenant.entity.SubscriptionCouponStatus;
import org.springframework.data.domain.Pageable;

/** CR-032 - plan trial coupons the OWNER creates and redeems for their own tenant. See SubscriptionCoupon for the full design note. */
public interface SubscriptionCouponService {

    SubscriptionCouponResponse create(SubscriptionCouponRequest request);

    SubscriptionCouponResponse update(Long id, SubscriptionCouponRequest request);

    PageResponse<SubscriptionCouponResponse> search(String search, SubscriptionCouponStatus status, Pageable pageable);

    void delete(Long id);

    SubscriptionCouponRedemptionResponse redeem(String code);
}

File: frontend/src/modules/settings/services/subscriptionCouponService.ts
import { apiDelete, apiGet, apiPost, apiPut } from '@/services/apiClient';
import type { PageResponse } from '@/shared/types/api';
import type {
  SubscriptionCouponRedemptionResponse, SubscriptionCouponRequest, SubscriptionCouponResponse,
  SubscriptionCouponStatus,
} from '../types';

/** Backend: tenant/controller/SubscriptionCouponController.java (CR-032) */
export const subscriptionCouponService = {
  search: (params: { search?: string; status?: SubscriptionCouponStatus; size?: number } = {}) =>
    apiGet<PageResponse<SubscriptionCouponResponse>>('/v1/subscription-coupons', { params }),

  create: (body: SubscriptionCouponRequest) =>
    apiPost<SubscriptionCouponResponse>('/v1/subscription-coupons', body),

  update: (id: number, body: SubscriptionCouponRequest) =>
    apiPut<SubscriptionCouponResponse>(`/v1/subscription-coupons/${id}`, body),

  remove: (id: number) => apiDelete(`/v1/subscription-coupons/${id}`),

  redeem: (code: string) =>
    apiPost<SubscriptionCouponRedemptionResponse>('/v1/subscription-coupons/redeem', { code }),
};
